link/src/main/java/com/stripe/android/link/ui/signup/SignUpViewModel.kt

```diff
@@ -0,0 +1,201 @@
+        }
+        linkEventsReporter.onSignupFlowPresented()
+    }
+
+    private suspend fun signUpEnabledListener() {
+        _state.flatMapLatest { state ->
+            combine(
+                flow = state.nameController.fieldState.map {
+                    if (requiresNameCollection) {
+                        it.isValid()
+                    } else {
+                        true
+                    }
+                },
+                flow2 = state.emailController.fieldState.map { it.isValid() },
+                flow3 = state.phoneNumberController.isComplete
+            ) { nameComplete, emailComplete, phoneComplete ->
+                nameComplete && emailComplete && phoneComplete
+            }
+        }.collectLatest { formValid ->
+            updateState { it.copy(signUpEnabled = formValid) }
+        }
+    }
+
+    private suspend fun emailListener() {
+        _state.flatMapLatest { state ->
+            state.emailController.formFieldValue.mapLatest { entry ->
+                entry.takeIf { it.isComplete }?.value
+            }
+        }.collectLatest { email ->
+            delay(LOOKUP_DEBOUNCE)
+            if (email != null) {
+                updateSignUpState(SignUpState.VerifyingEmail)
+                lookupConsumerEmail(email)
+            } else {
+                updateSignUpState(SignUpState.InputtingPrimaryField)
+            }
+        }
+    }
+
+    fun onSignUpClick() {
+        clearError()
+        viewModelScope.launch {
+            val state = _state.value
+            linkAccountManager.signUp(
+                email = state.emailController.fieldValue.value,
+                phone = state.phoneNumberController.fieldValue.value,
+                country = state.phoneNumberController.getCountryCode(),
+                name = state.nameController.fieldValue.value,
+                consentAction = SignUpConsentAction.Implied
+            ).fold(
+                onSuccess = {
+                    onAccountFetched(it)
+                    linkEventsReporter.onSignupCompleted()
+                },
+                onFailure = {
+                    onError(it)
+                    linkEventsReporter.onSignupFailure(error = it)
+                }
+            )
+        }
+    }
+
+    private fun onAccountFetched(linkAccount: LinkAccount?) {
+        if (linkAccount?.isVerified == true) {
+            navController?.popBackStack(LinkScreen.Wallet.route, inclusive = false)
+        } else {
+            navController?.navigate(LinkScreen.Verification.route)
+            // The sign up screen stays in the back stack.
+            // Clean up the state in case the user comes back.
+            _state.value.emailController.onValueChange("")
+        }
+    }
+
+    private suspend fun lookupConsumerEmail(email: String) {
+        clearError()
+        linkAccountManager.lookupConsumer(email).fold(
+            onSuccess = {
+                if (it != null) {
+                    onAccountFetched(it)
```

Comment:
should this be done before we even get to this screen? It seems like this would always navigate away from the sign up screen, so I wonder if we should do it while loading or something instead

Reply:
This will be checked on the LinkVM. The user will not get to this screen if we find an account. The first lookup  here will return null (most likely) since we already checked beforehand. 

This VM also listens to email changes. If the lookup is successful after an email change, we would navigate to the next screen as well. This is not likely and I'm certain it won't be the case since we don't have persistent log in anymore. I decided to keep in case we change our mind to add persistence later on. The old link implementation had persistence and did a lookup anytime the email changed.

Reply:
Gotcha, it sounds like this check is unnecessary right now then, is that right? If we add persistence back, we should add this code snippet back. But I think we should avoid adding this unless it's needed.